src/h5web/visualizations/line/LineVis.tsx

```diff
@@ -27,36 +33,66 @@ function LineVis(props: Props): ReactElement {
     curveType = CurveType.LineOnly,
     showGrid = true,
     scaleType = ScaleType.Linear,
+    abscissas,
   } = props;
 
-  const indexDomain = extendDomain([0, dataArray.size - 1], 0.01);
-  const dataDomain = useMemo(() => {
-    return domain
-      ? extendDomain(domain, 0.05, scaleType === ScaleType.Log)
+  const abscissaValues = abscissas || range(dataArray.size);
+  const abscissaToIndex = getValueToIndexScale(
+    abscissaValues.map(
+      (x, i) =>
+        abscissaValues[i - 1] + (abscissaValues[i] - abscissaValues[i - 1]) / 2
+    )
+  ); // Shifting index change threshold to midpoints
+  const abscissaDomain = useMemo(() => {
+    const rawDomain = getDataDomain(abscissaValues);
+    return rawDomain ? extendDomain(rawDomain, 0.01) : DEFAULT_DOMAIN;
+  }, [abscissaValues]);
+
+  const ordinateValues = dataArray.data as number[];
+  const ordinateDomain = useMemo(() => {
+    const rawDomain = getDataDomain(ordinateValues, scaleType);
+    return rawDomain
+      ? extendDomain(rawDomain, 0.05, scaleType === ScaleType.Log)
       : DEFAULT_DOMAIN;
-  }, [scaleType, domain]);
+  }, [ordinateValues, scaleType]);
+
+  if (abscissaValues.length !== dataArray.size) {
+    return (
+      <p>{`Abscissas size (${abscissaValues.length}) does not match data length (${dataArray.size})`}</p>
+    );
+  }
 
   return (
     <div className={styles.root}>
       <VisCanvas
-        abscissaConfig={{ indexDomain, showGrid }}
-        ordinateConfig={{ dataDomain, showGrid, scaleType }}
+        abscissaConfig={{
+          domain: abscissaDomain,
+          showGrid,
+          isIndexAxis: !abscissas,
+        }}
+        ordinateConfig={{
+          domain: domain || ordinateDomain,
+          showGrid,
+          scaleType,
+        }}
       >
         <TooltipMesh
-          formatIndex={([x]) => `x=${Math.round(x)}`}
+          formatIndex={([x]) =>
+            `x=${format('0')(abscissaValues[abscissaToIndex(x)])}`
+          }
           formatValue={([x]) => {
-            const value = dataArray.get(Math.round(x));
+            const index = abscissaToIndex(x);
+            const value = ordinateValues[index];
             return value !== undefined ? format('.3f')(value) : undefined;
```

Comment:
I wouldn't worry with the `index` constant here. Also, can the `value` still be `undefined`?

Reply:
:+1: for `index`

As for `value`, given that we extend the domain in `x`, we could imagine having a value under the mouse that doesn't correspond to an index, leading to `undefined` for `value`.

Reply:
Ah right, forgot about the domain extension.